File: src/main/java/xyz/nkomarn/composter/network/NetworkController.java
package xyz.nkomarn.composter.network;

import io.netty.bootstrap.ServerBootstrap;
import io.netty.channel.EventLoopGroup;
import io.netty.channel.ServerChannel;
import io.netty.channel.epoll.Epoll;
import io.netty.channel.epoll.EpollEventLoopGroup;
import io.netty.channel.epoll.EpollServerSocketChannel;
import io.netty.channel.nio.NioEventLoopGroup;
import io.netty.channel.socket.nio.NioServerSocketChannel;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import xyz.nkomarn.composter.network.pipeline.InboundInitializer;
import xyz.nkomarn.composter.server.MinecraftServer;

import static io.netty.channel.ChannelOption.TCP_NODELAY;

public class NetworkController {

    private static final Logger LOGGER = LoggerFactory.getLogger("Network");
    private final ServerBootstrap bootstrap;

    public NetworkController(MinecraftServer server) {
        this.bootstrap = new ServerBootstrap()
                .group(createGroup())
                .channel(channelClass())
                .childOption(TCP_NODELAY, true)
                .childHandler(new InboundInitializer(server));
    }

    public void bind(int port) throws InterruptedException {
        var future = bootstrap.bind(port).sync();

        if (future.isSuccess()) {
            LOGGER.info("Bound to port {} and accepting connections.", port);
        }

        future.channel().closeFuture().sync();
    }

    private EventLoopGroup createGroup() {
        return Epoll.isAvailable()
                ? new EpollEventLoopGroup()
                : new NioEventLoopGroup();
    }

    private Class<? extends ServerChannel> channelClass() {
        return Epoll.isAvailable()
                ? EpollServerSocketChannel.class
                : NioServerSocketChannel.class;
    }
}

File: src/main/kotlin/kyta/composter/network/NetworkController.kt
package kyta.composter.network

import io.netty.bootstrap.ServerBootstrap
import io.netty.channel.ChannelOption
import io.netty.channel.EventLoopGroup
import io.netty.channel.ServerChannel
import io.netty.channel.epoll.Epoll
import io.netty.channel.epoll.EpollEventLoopGroup
import io.netty.channel.epoll.EpollServerSocketChannel
import io.netty.channel.nio.NioEventLoopGroup
import io.netty.channel.socket.nio.NioServerSocketChannel
import kyta.composter.network.pipeline.InboundConnectionInitializer
import org.slf4j.LoggerFactory
import kyta.composter.server.MinecraftServer
import kotlin.reflect.KClass

class NetworkController(private val server: MinecraftServer) {
    private val logger = LoggerFactory.getLogger("net")

    @Throws(InterruptedException::class)
    fun bind(port: Int) {
        val future = ServerBootstrap()
            .group(createGroup())
            .channel(channelClass().java)
            .childOption(ChannelOption.TCP_NODELAY, true)
            .childHandler(InboundConnectionInitializer(server, logger))
            .bind(port)
            .sync()

        if (future.isSuccess) {
            logger.info("bound to port {} and accepting connections.", port)
        }

        future.channel().closeFuture().sync()
    }

    private fun createGroup(): EventLoopGroup {
        return if (Epoll.isAvailable()) EpollEventLoopGroup() else NioEventLoopGroup()
    }

    private fun channelClass(): KClass<out ServerChannel> {
        return if (Epoll.isAvailable()) EpollServerSocketChannel::class else NioServerSocketChannel::class
    }
}